src/trs/rewrite.rs

```diff
@@ -0,0 +1,128 @@
+//! Hindley-Milner Typing for First-Order Term Rewriting Systems (no abstraction)
+//!
+//! Much thanks to:
+//! - https://github.com/rob-smallshire/hindley-milner-python
+//! - https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system
+//! - (TAPL; Pierce, 2002, ch. 22)
+
+use polytype::{Context, TypeSchema};
+use rand::Rng;
+use std::f64::NEG_INFINITY;
+use std::fmt;
+use term_rewriting as trs;
+
+use super::trace::Trace;
```

Comment:
We decided to move `Trace` to [term_rewriting](https://github.com/joshrule/term-rewriting-rs). Do you think it makes sense to make that move in this PR or a future one?

Reply:
We should save it for a future one — it'd make this PR even more of a behemoth of changes.